### PR TITLE
Add S3 storage class configuration option

### PR DESCRIPTION
**Pull request by patf**
A new environment variable called `WALG_S3_STORAGE_CLASS` allows users to set the storage class used when wal-g writes objects to S3. This allows usage of the Infrequent Access or Reduced
Redundancy features of S3.

In keeping with current behaviour, new objects are created with the "STANDARD" storage class if the variable is not set.

Note: The empty json file created at the end of base backups is created using the "STANDARD" storage class regardless of this setting. Infrequent Access has a 128KB minimum charge, so using it for a 2-byte file would make little sense.

### CODE REVIEW

**Comment by fdr** [comment]
Seems reasonable. Two questions:

1. Testing, to ensure that setting environment variables has the right effect. We're discussing this in #16 

2. What's the error handling if one types in an invalid storage class here?

**Comment by patf** [comment]
Thanks for the feedback!

> Testing, to ensure that setting environment variables has the right effect. We're discussing this in #16

This is tricky to test beyond the existing configuration test. My current idea is to introduce a new unexported function in `upload.go` which takes a path, an `io.Reader` and a `TarUploader` and returns an `UploadInput`. This would be easy to test against and would be slightly more DRY.

One obstacle is that tests are currently in their own package (`walg_test`), meaning I would not be able to call the function directly. IIRC in idiomatic Go, the tests should be in the same package as the code they're testing. Is there a reason this is done differently in wal-g?

I'd be happy to change the package for all existing tests (probably in a separate PR, which this one would build on). Worth noting that this would be a relatively big diff since all the `walg.` references in all tests would have to be replaced. Alternatively, I could add a new `_test` file in the `walg` package which would do nothing but export the new function so it can be tested.

If you can think of a better way of testing this, do let me know!

> What's the error handling if one types in an invalid storage class here?

S3 seems to silently fall back to STANDARD if an invalid value is used (on their end, not in the SDK). The SDK doesn't provide a validation function, shall we hardcode the storage classes?

**Comment by fdr** [comment]
The matter of idiom is, as I recall, on being able to avoid package-private symbols. I don't think wal-g has a systematic approach to this that I know of, though at least enough must be exposed to allow `cmd/wal-g` and the development tools in `test_tools` to link properly. I don't think there's a need to change it for now.

However, if you need access to private symbols, feel free to modify a particular test file to be in-package (if your test would seem to assort well alongside some others), or make a new file if there is no existing test file for which your tests would seem to fit in with.

> S3 seems to silently fall back to STANDARD if an invalid value is used (on their end, not in the SDK). The SDK doesn't provide a validation function, shall we hardcode the storage classes?

I think we can abide by the S3 silent-fallback behavior unless it proves particularly problematic. That way there's no project turnaround time for new storage classes.

**Comment by patf** [comment]
The last commit moves the `UploadInput` code to a new `createUploadInput` function and adds tests for the `StorageClass` that's actually used in the `UploadInput`. Ready for re-review.

File: upload_storage_class_test.go [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+	}
+
+	// Test STANDARD_IA storage class
+	tu = NewTarUploader(nil, "bucket", "server", "region", 1, float64(1))
+	tu.StorageClass = "STANDARD_IA"
+	input = tu.createUploadInput("path", nil)
+	if *input.StorageClass != "STANDARD_IA" {
+		t.Errorf("upload: UploadInput field 'StorageClass' expected %s but got %s", "STANDARD_IA", *input.StorageClass)
+	}
+}
```
[the remark this answers is not in the document]
Comment (patf) [reply upload_storage_class_test.go]: newline added - the other files were fine.

**Comment by fdr** [comment]
Squashed and merged in 0c6fa43c2912c86e22bfe3c19502f9147361d980

**Pull request closed without merging**